pkg/sysfs/system.go

```diff
@@ -732,6 +736,118 @@ func (c *cpu) SetFrequencyLimits(min, max uint64) error {
 	return nil
 }
 
+// CacheCount returns the number of caches for this CPU.
+func (c *cpu) CacheCount() int {
+	return len(c.caches)
+}
+
+// GetCaches returns the caches for this CPU.
+func (c *cpu) GetCaches() []*Cache {
+	caches := make([]*Cache, 0, len(c.caches))
+	copy(caches, c.caches)
+	return caches
+}
+
+// GetCachesByLevel returns the caches of the given level for this CPU.
+func (c *cpu) GetCachesByLevel(level int) []*Cache {
+	var caches []*Cache
+
+	for _, cache := range c.caches {
+		if cache.level == level {
+			caches = append(caches, cache)
+		}
+	}
+
+	return caches
+}
+
+// GetCacheByIndex returns the cache of the given index for this CPU.
+func (c *cpu) GetCacheByIndex(idx int) *Cache {
+	if 0 <= idx && idx < len(c.caches) {
+		return c.caches[idx]
+	}
+	return nil
+}
+
+// GetLastLevelCaches returns the last level caches for this CPU.
+func (c *cpu) GetLastLevelCaches() []*Cache {
+	var (
+		caches []*Cache
+		level  = -1
+	)
+
+	for idx := len(c.caches) - 1; idx >= 0; idx-- {
+		cch := c.caches[idx]
+		if level == -1 {
+			level = cch.level
+		}
+		if level != cch.level {
+			break
+		}
+		caches = append(caches, cch)
```

Comment:
This loop somehow looks haphazard. We always assume `c.caches` to be sorted? Is c.caches[idx].level sometimes `-1`?

Reply:
It's set to -1 before the loop then, if there are any caches, it gets set to the `level` of the `index0` cache, to collect both types of caches at that level if the cache is not unified (which it usually is not).

Reply:
But you were right, it haphazard. Not only that, it was also broken. Fixed and rewrote it to be hopefully more easy to follow.